validation/src/validation_service/mod.rs

```diff
@@ -248,32 +262,36 @@ fn signing_key(validators: &[ValidatorId], keystore: &KeyStorePtr) -> Option<Arc
 }
 
 /// Constructs parachain-agreement instances.
-pub(crate) struct ParachainValidationInstances<C, N, P, SP> {
+pub(crate) struct ParachainValidationInstances<N, P, SP, CFB, CF, CFF, Error> {
 	/// The client instance.
 	client: Arc<P>,
 	/// The backing network handle.
 	network: N,
-	/// Parachain collators.
-	collators: C,
 	/// handle to spawner
 	spawner: SP,
 	/// Store for extrinsic data.
 	availability_store: AvailabilityStore,
 	/// Live agreements. Maps relay chain parent hashes to attestation
 	/// instances.
 	live_instances: HashMap<Hash, ValidationInstanceHandle>,
+	/// Used to fetch a collation.
+	collation_fetch_builder: CFB,
+	_phantom: PhantomData<(CF, CFF, Error)>
 }
 
-impl<C, N, P, SP> ParachainValidationInstances<C, N, P, SP> where
-	C: Collators + Send + Unpin + 'static + Sync,
+impl<N, P, SP, CFB, CF, CFF, Error> ParachainValidationInstances<N, P, SP, CFB, CF, CFF, Error> where
 	N: Network,
-	P: ProvideRuntimeApi<Block> + HeaderBackend<Block> + BlockBackend<Block> + Send + Sync + 'static,
-	P::Api: ParachainHost<Block> + BlockBuilderApi<Block> + ApiExt<Block, Error = sp_blockchain::Error>,
-	C::Collation: Send + Unpin + 'static,
-	N::TableRouter: Send + 'static,
+	N::Error: 'static,
+	P: ProvideRuntimeApi<Block> + Send + Sync + 'static,
+	P::Api: ParachainHost<Block, Error = sp_blockchain::Error>,
+	N::TableRouter: Send + 'static + Sync,
 	<N::TableRouter as TableRouter>::SendLocalCollation: Send,
 	N::BuildTableRouter: Unpin + Send + 'static,
 	SP: Spawn + Send + 'static,
+	CFB: Fn() -> CF + Send,
+	CF: FnOnce(ParaId, Hash, Arc<P>, Option<u64>, usize) -> CFF + Send + 'static,
```

Comment:
This code is less readable than before. It's not clear what all of those parameters are from the signature